src/riak_kv_qry_compiler.erl

```diff
@@ -436,7 +443,24 @@ col_index_and_type_of(Fields, ColumnName) ->
             {Position, Type}
     end.
 
-expand_where(Where, #key_v1{ast = PAST}, MaxSubQueries) ->
+%%
+-spec expand_where(filter(), #key_v1{}, integer()) ->
+        [where_props()] | {error, any()}.
+expand_where(Where, PartitionKey, MaxSubQueries) ->
+    case find_quanta_function_in_key(PartitionKey) of
+        [{[QField], QSize, QUnit}] ->
+            hash_timestamp_to_quanta(QField, QSize, QUnit, MaxSubQueries, Where);
+        [] ->
+            [Where]
+    end.
+
+%%
+find_quanta_function_in_key(#key_v1{ ast = PKAST }) ->
+    [{X, Y, Z} || #hash_fn_v1{mod = riak_ql_quanta, fn = quantum,
+                              args = [#param_v1{name = X}, Y, Z]} <- PKAST].
+
+%%
+hash_timestamp_to_quanta(QField, QSize, QUnit, MaxSubQueries, Where) ->
     GetMaxMinFun = fun({startkey, List}, {_S, E}) ->
                            {element(3, lists:last(List)), E};
```

Comment:
Again, a little opaque on what's being dug out - it seems like it's the value of the last element of the composite key. Maybe a helper function would be useful for digging out name / type / value from the startkeys, otherwise you'd have to wrap in a begin/end block which would probably be worse. 

Your choice - at minimum a comment explaining we're digging out the min/max timestamps to compute the quanta on from the start key.